### PR TITLE
Профиль пользователя

### PR DESCRIPTION
**Pull request by NikitaGoroshko7**
https://trello.com/c/ySpS3hkh/38-%D0%BF%D1%80%D0%BE%D1%84%D0%B8%D0%BB%D1%8C-%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D1%82%D0%B5%D0%BB%D1%8F

Has been added ProfileController
Has been created migrations and added extra properties for user
Has been added Views: Index, Settings, CreatedProjects and Balance
Has been added ViewModel for user setting's
Has been added a modal window for exit from profile and for delete profile
Has been added style and script for the modal window
Has been added folder Avatars for store user avatar's
Has been changed import.cs, was added using for work with files

### CODE REVIEW

File: CrowdFundingApplication/CrowdFundingApplication.csproj [hunk 1]
```diff
@@ -15,4 +15,8 @@
     </PackageReference>
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="wwwroot\Avatars\" />
```
Comment (GalayM) [inline CrowdFundingApplication/CrowdFundingApplication.csproj:19]: This folder is not accessible via a link without an authorization?

File: CrowdFundingApplication/wwwroot/js/ModalWindow.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        modal = document.getElementById(e.target.dataset.id);
+        openModal(modal);
+    } else if (e.target.className === "modal-close") {
+        closeModal(modal);
+    } else {
+        return;
+    }
+});
+
+const openModal = (modal) => {
```
Comment (GalayM) [inline CrowdFundingApplication/wwwroot/js/ModalWindow.js]: https://getbootstrap.com/docs/4.0/components/modal/

File: CrowdFundingApplication/wwwroot/js/ModalWindow.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿let modal;
+document.addEventListener("click", (e) => {
```
Comment (GalayM) [inline CrowdFundingApplication/wwwroot/js/ModalWindow.js]: Using jquery

File: CrowdFundingApplication/Views/Profile/Settings.cshtml [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                                <div class="col">
+                                    <a class="btn btn-success btn-block" asp-area="" asp-controller="Profile" asp-action="Settings">Нет</a>
+                                    <br />
+                                </div>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+
+                <script>
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Profile/Settings.cshtml]: move to js file or js-section https://docs.microsoft.com/en-us/aspnet/core/mvc/views/layout?view=aspnetcore-6.0

File: CrowdFundingApplication/Views/Profile/Settings.cshtml [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                                    <label class="form-label" for="lastName">Фотография</label>
+                                </div>
+                                <img id="output" class="rounded-circle" style="width: 150px; height: 160px;">
+                                <div class="row mt-1">
+                                    <div class="col">
+                                        <input type="file" name="Image" id="file" onchange="loadFile(event)" style="display: none;" />
+                                        <p><label for="file" name="Image" style="cursor: pointer;">Изменить</label></p>
+                                    </div>
+                                    <div class="col">
+                                        <p><label for="delete" style="cursor: pointer;" onclick="deleteFile()">Удалить</label></p>
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Profile/Settings.cshtml:147]: Better link event via id + jquery

File: CrowdFundingApplication/Views/Profile/Balance.cshtml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                <li class="nav-item">
+                    <a class="nav-link active" style="color: black;" aria-current="page" href="#">Управление балансом</a>
+                </li>
+            </ul>
+        </div>
+    </div>
+
+    <div class="row mt-1">
+        <div class="col">
+            <div class="card shadow-2-strong card-registration" style="border-radius: 5px;">
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Profile/Balance.cshtml:27]: click should not be in html

File: CrowdFundingApplication/Models/User.cs [hunk 7]
```diff
@@ -2,6 +2,12 @@
 {
     public class User : IdentityUser
     {
-
+        public string Avatar { get; set; }
+        public string Description { get; set; }
+        public int Balance { get; set; }
+        public int Age { get; set; }
+        public string Sex { get; set; }
```
Comment (GalayM) [inline CrowdFundingApplication/Models/User.cs]: Using enum

File: CrowdFundingApplication/Models/User.cs [hunk 8]
```diff
@@ -2,6 +2,12 @@
 {
     public class User : IdentityUser
     {
-
+        public string Avatar { get; set; }
+        public string Description { get; set; }
+        public int Balance { get; set; }
```
Comment (GalayM) [inline CrowdFundingApplication/Models/User.cs]: If it many - better decimal

File: CrowdFundingApplication/Models/User.cs [hunk 9]
```diff
@@ -2,6 +2,12 @@
 {
     public class User : IdentityUser
     {
-
+        public string Avatar { get; set; }
```
Comment (GalayM) [inline CrowdFundingApplication/Models/User.cs]: Avatar -> AvatarFileName

File: CrowdFundingApplication/Models/EditProfileViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    [DataType(DataType.Password)]
+    public string NewPassword { get; set; }
+
+    [DataType(DataType.Password)]
+    [Compare("NewPassword", ErrorMessage = "Пароль введен неверно")]
+    public string ConfirmPassword { get; set; }
+
+    [StringLength(100, MinimumLength = 10, ErrorMessage = "Длина строки должна быть от 10 до 100 символов !")]
+    public string Description { get; set; }
+
```
Comment (GalayM) [inline CrowdFundingApplication/Models/EditProfileViewModel.cs]: ???

File: CrowdFundingApplication/Models/EditProfileViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
+﻿namespace CrowdFundingApplication.Models;
+
+public class EditProfileViewModel
+{
+    [EmailAddress (ErrorMessage = "Неверный email адрес !")]
```
Comment (GalayM) [inline CrowdFundingApplication/Models/EditProfileViewModel.cs:5]: Неверный email адрес ! -> Неверный email адрес! 
In all cases

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        var user = await _userManager.FindByNameAsync(HttpContext.User.Identity.Name);
+        return View(user);
+    }
+
+    [HttpGet]
+    public IActionResult Settings() => View();
+
+    [HttpPost]
+    public async Task<IActionResult> Settings(EditProfileViewModel model, IFormFile Image)
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: It is work? Did you test?

File: CrowdFundingApplication/Views/Profile/Index.cshtml [hunk 13]
```diff
@@ -0,0 +1,146 @@
+﻿@model CrowdFundingApplication.Models.User
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Profile/Index.cshtml]: You cannot use domain models for mapping. In all cases.

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        }
+        return uniqueFileName;
+    } 
+
+    public async Task<IActionResult> DeleteProfile()
+    {
+        var user = await _userManager.FindByNameAsync(HttpContext.User.Identity.Name);
+        await _userManager.DeleteAsync(user);
+        return RedirectToAction("MainPage", "Home");
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: Logout?

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private string UploadedFile(IFormFile image)
+    {
+        string uniqueFileName = null;
+
+        if (image is not null)
+        {
+            string uploadsFolder = Path.Combine(_webHostEnvironment.WebRootPath, "Avatars");
+            uniqueFileName = Guid.NewGuid().ToString() + "_" + image.FileName;
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: We discussed it!

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        return View();
+    }
+
+    private string UploadedFile(IFormFile image)
+    {
+        string uniqueFileName = null;
+
+        if (image is not null)
+        {
+            string uploadsFolder = Path.Combine(_webHostEnvironment.WebRootPath, "Avatars");
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: string -> var

There must be 1 style or we always write styles or always var

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        if (ModelState.IsValid)
+        {
+            var user = await _userManager.FindByNameAsync(HttpContext.User.Identity.Name);
+            if (user is null)
+            {
+                return View(model);
+            }
+
+            //update dates for user
+            if (model.Age is not null) user.Age = (int)model.Age;
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs:35]: if Age = null throw exeprion

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+            if (model.Email is not null) user.Email = model.Email;
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: You can't just change the email - look at this table in the database and you will understand what I mean.

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+            if (model.NumberOfPhone is not null) user.PhoneNumber = model.NumberOfPhone;
+            if (model.Sex is not null) user.Sex = model.Sex;
+            if (model.Country is not null) user.Country = model.Country;
+            if (model.City is not null) user.City = model.City;
+            if (model.Description is not null) user.Description = model.Description;
+            if (Image is not null) user.Avatar = UploadedFile(Image);
+
+            if (model.NewPassword is not null)//update password for user
+            {
+                var _passwordValidator = HttpContext.RequestServices.GetService(typeof(IPasswordValidator<User>)) as IPasswordValidator<User>;
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: Use DI via constructor

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+                var _passwordHasher = HttpContext.RequestServices.GetService(typeof(IPasswordHasher<User>)) as IPasswordHasher<User>;
+                IdentityResult result = await _passwordValidator.ValidateAsync(_userManager, user, model.NewPassword);
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: http://dotnet-tutorials.net/identity/update-user-password

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+    public ProfileController(UserManager<User> userManager, IWebHostEnvironment webHostEnvironment)
+    {
+        _userManager = userManager;
+        _webHostEnvironment = webHostEnvironment;
+    }
+
+    public async Task<IActionResult> Index()
+    {
+        var user = await _userManager.FindByNameAsync(HttpContext.User.Identity.Name);
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: HttpContext.User.Identity.Name => User.Identity.Name

File: CrowdFundingApplication/CrowdFundingApplication.csproj (hunk 1 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/CrowdFundingApplication.csproj:19]: Yes, not accessible
This folder just store user avatars

File: CrowdFundingApplication/wwwroot/js/ModalWindow.js (hunk 2 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/wwwroot/js/ModalWindow.js]: In my case component bootstrap modal window did not work that's why I was using my custom modal window

File: CrowdFundingApplication/Views/Profile/Balance.cshtml (hunk 6 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Profile/Balance.cshtml:27]: I did not understand which exactly click?

File: CrowdFundingApplication/Models/User.cs (hunk 7 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/User.cs]: But I need all this properties for user

File: CrowdFundingApplication/Models/EditProfileViewModel.cs (hunk 11 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/EditProfileViewModel.cs:5]: If I will not use this attribute then all my issues will be in English, that's why I used it

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 12 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Yes it works

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 14 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: No, this is action which is delete user from database

File: CrowdFundingApplication/CrowdFundingApplication.csproj (hunk 1 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/CrowdFundingApplication.csproj:19]: Has been edited

File: CrowdFundingApplication/wwwroot/js/ModalWindow.js (hunk 2 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/wwwroot/js/ModalWindow.js]: Has been edited

File: CrowdFundingApplication/wwwroot/js/ModalWindow.js (hunk 3 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/wwwroot/js/ModalWindow.js]: Has been edited

File: CrowdFundingApplication/Views/Profile/Settings.cshtml (hunk 4 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Profile/Settings.cshtml]: Has been edited

File: CrowdFundingApplication/Views/Profile/Settings.cshtml (hunk 5 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Profile/Settings.cshtml:147]: Has been edited

File: CrowdFundingApplication/Views/Profile/Balance.cshtml (hunk 6 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Profile/Balance.cshtml:27]: Has been edited

File: CrowdFundingApplication/Models/User.cs (hunk 7 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/User.cs]: Has been edited

File: CrowdFundingApplication/Models/User.cs (hunk 8 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/User.cs]: Has been edited

File: CrowdFundingApplication/Models/User.cs (hunk 9 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/User.cs]: Has been edited

File: CrowdFundingApplication/Models/EditProfileViewModel.cs (hunk 10 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/EditProfileViewModel.cs]: Has been edited

File: CrowdFundingApplication/Models/EditProfileViewModel.cs (hunk 11 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/EditProfileViewModel.cs:5]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 12 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 14 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 16 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 17 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs:35]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 18 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 19 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 20 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 21 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 15 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Views/Profile/Index.cshtml (hunk 13 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Profile/Index.cshtml]: Has been edited

File: CrowdFundingApplication/Models/UserViewModel.cs [hunk 22]
```diff
@@ -0,0 +1,15 @@
+﻿namespace CrowdFundingApplication.Models;
+
+public class UserViewModel
+{
+    public string UserName { get; set; }
+    public string AvatarFileName { get; set; }
+    public string Email { get; set; }
+    public string PhoneNumber { get; set; }
+    public int Age { get; set; }
```
Comment (GalayM) [inline CrowdFundingApplication/Models/UserViewModel.cs:9]: int -> Sex

Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Models/UserViewModel.cs:9]: Has been edited

**Review by GalayM — APPROVED** [review]

**Pull request merged**